python/lsst/daf/butler/_butler.py

```diff
@@ -1755,67 +1763,94 @@ def transfer_from(self, source_butler: Butler, source_refs: Iterable[DatasetRef]
 
         # Importing requires that we group the refs by dataset type and run
         # before doing the import.
+        accepted_source_refs = []
         grouped_refs = defaultdict(list)
         grouped_indices = defaultdict(list)
-        for i, ref in enumerate(source_refs):
+        i = 0
+        for ref in source_refs:
+            # In some situations the datastore artifact may be missing
+            # and we do not want that registry entry to be imported.
+            # Asking datastore is not sufficient, the records may have been
+            # purged, we have to ask for the (predicted) URI and check
+            # existence explicitly. Execution butler is set up exactly like
+            # this with no datastore records.
+            if skip_missing:
+                if not source_butler.datastore.exists(ref):
+                    continue
+
+            accepted_source_refs.append(ref)
             grouped_refs[ref.datasetType, ref.run].append(ref)
             grouped_indices[ref.datasetType, ref.run].append(i)
 
+            # Update counter since enumerate cannot be used since we
+            # are filtering the refs.
+            i += 1
```

Comment:
I think you can make logic somewhat simpler and keep `enumerate` if you filter separately:

    if skip_missing:
        source_refs = [ref for ref in source_refs if source_butler.datastore.exists(ref)]